Add documentation for reviewers and a checklist for PRs

smoia opened the pull request:
Closes none, depends on #313.

## Proposed Changes

  - Update the PR template to have a checklist of steps that should be completed before promoting a draft PR to "Open PR" (Or before creating one) and more comments.
  - Add a documentation section on reviewing
  - Add a documentation section on Main Reviewers

smoia commented:
I know we didn't solve the issues with `auto` yet, but I'm going to unpause this already.
@physiopy/phys2bids anyone wants to have a look?

eurunuela commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,7 +1,36 @@
+<!-- Write all the issues that are linked to this one. -->
+<!-- If this PR is enough to close them you can write something like "Closes #314 and closes #42" -->
+<!-- If it just reference them without solving them, you can add something like "References #112" -->
```
```suggestion
<!-- If you just want to reference them without closing them, you can add something like "References #112" -->
```

eurunuela commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,7 +1,36 @@
 Closes #
 
-## Proposed Changes
+<!-- Add here a short description of the PR content -->
```
```suggestion
<!-- Add a short description of the PR content here-->
```

eurunuela commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,7 +1,36 @@
+- [ ] `documentation` (no version update)
+
+## Checklist before review
+<!-- If this section is not clear, please read this part of the dodcs: https://phys2bids.readthedocs.io/en/latest/contributorfile.html#pr -->
```
```suggestion
<!-- If this section is not clear, please read this part of the docs: https://phys2bids.readthedocs.io/en/latest/contributorfile.html#pr -->
```

eurunuela commented on docs/contributing.rst:
```diff
@@ -9,12 +9,11 @@ First of all: thank you!
-This will let you run the program with the latest modification, without requiring you to re-install it every time.
+1. Have a look at the `contributor guide <contributorfile.html>`_ page as well as the `code of conduct <conduct.html>`_.
+2. Make sure that you have a GitHub account. You can set up a `free GitHub account <https://github.com/>`_; here are some `instructions <https://help.github.com/articles/signing-up-for-a-new-github-account>`_.
+3. If you intend to contribute code and/or use the ``physiopy`` packages in any way, check that you have ``git`` and ``pip`` installed on your system. Then install ``phys2bids`` as a developer. This will let you run the program with the latest modification, without requiring to re-install it every time.
```
```suggestion
3. If you intend to contribute code and/or use the ``physiopy`` packages in any way, check that you have ``git`` and ``pip`` installed on your system. Then install ``phys2bids`` as a developer. This will let you run the program with the latest modifications, without requiring to re-install it every time.
```

eurunuela reviewed (changes requested):
I've some tiny suggestions.

RayStick commented:
Good initiative - this will be helpful. 

Do you think including something about the timeline of a review would be needed? (Sorry if I've missed this somewhere)

I.e. "I know this PR needs to be merged before we can work on X", or "it would be great to get this done by X day". Then a reviewer that is assigned to a few PRs might know how best to spend their time, or they say they don't think they can get it done in that time frame (yes, I might be thinking personally here!). I guess that's kind of what the milestones are for, right?

RayStick commented:
I suggested some changes to your branch @smoia. Other than that, LGTM

RayStick reviewed (changes requested):
I suggested some changes to your branch @smoia. Other than that, LGTM

tsalo commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,7 +1,36 @@
+- [ ] \[Code or tests only\] I ran and passed tests locally.
+- [ ] \[Documentation only\] I built the docs locally.
+- [ ] My contribution is harmonious with the rest of the code: I'm not introducing repetitions and I respect the general style.
+<!-- "General style" indicate those non-mandatory rules such as `'` instead of `"` and these sort of things -->
```
You might need to change that based on #324 😆

tsalo commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,7 +1,36 @@
+- [ ] \[Documentation only\] I built the docs locally.
+- [ ] My contribution is harmonious with the rest of the code: I'm not introducing repetitions and I respect the general style.
+<!-- "General style" indicate those non-mandatory rules such as `'` instead of `"` and these sort of things -->
+- [ ] My code respects the adopted style, especially lintering conventions.
```
```suggestion
- [ ] My code respects the adopted style, especially linting conventions.
```
(or `linter` I suppose, if the point is about the settings for the linter)

tsalo commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,7 +1,36 @@
+- [ ] The title of this PR is explicative on its own, enough to be understood as part of a changelog.
+- [ ] I added or indicated the right labels.
+<!-- If you want, you can ask for reviews on a draft PR -->
+- [ ] Please, comment my PR while it's a draft and give me feedback on the development!
```
```suggestion
- [ ] Please, comment on my PR while it's a draft and give me feedback on the development!
```

tsalo commented on docs/contributorfile.rst:
```diff
@@ -132,75 +156,143 @@ Issues labels
+Reviewing PRs is a time consuming task, that can be stressful for both the reviewer and the author. Avoiding wasting time and the need of little fixes - such as fixing grammar mistakes and typos, styling code, or adopting conventions - is a good start for a successful (and quick) review. Before graduating a Draft PR to a PR ready for review, please check that:
+
+- You did all you wanted to include in your PR. If at a later stage you realise something is missing and it's not a minor thing, you will need to open a new PR.
+- If your contribution contains code or tests, you run and passed all the tests locally with `pytest`.
```
```suggestion
- If your contribution contains code or tests, you ran and passed all of the tests locally with `pytest`.
```

smoia commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,7 +1,36 @@
+- [ ] \[Code or tests only\] I wrote/updated the necessary docstrings.
+- [ ] \[Code or tests only\] I ran and passed tests locally.
+- [ ] \[Documentation only\] I built the docs locally.
+- [ ] My contribution is harmonious with the rest of the code: I'm not introducing repetitions and I respect the general style.
```
```suggestion
- [ ] My contribution is harmonious with the rest of the code: I'm not introducing repetitions.
```

smoia commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,7 +1,36 @@
+- [ ] \[Code or tests only\] I ran and passed tests locally.
+- [ ] \[Documentation only\] I built the docs locally.
+- [ ] My contribution is harmonious with the rest of the code: I'm not introducing repetitions.
+<!-- "General style" indicate those non-mandatory rules such as `'` instead of `"` and these sort of things -->
```
```suggestion

```

tsalo commented:
One thing I didn't see (but might have missed) was something about scope. I personally have a habit of letting my PRs sprawl. Stuff like fixing typos outside of the code being targeted by the PR and such should probably be explicitly warned against. Apologies if this is in there, but I missed it.

smoia commented:
I went through the changes, accepted all of them (apologies @RayStick , I missed the PR to my branch!) and added a sentence to invite contributors to narrow the scope of PRs considering @tsalo's point above.
Let me know if there is anything more to do!

RayStick reviewed (approved):
Great. When others review, I will do all the final checks as main reviewer, and merge if it all looks good.

tsalo commented on docs/contributorfile.rst:
```diff
@@ -132,75 +156,144 @@ Issues labels
+
+- Takes care of the reviewing process of the PR, in particular:
+    - Invites the reviewers to finish their review in a relatively short time.
+    - Checks that this document was respected in all its part, especially the part about `Pull Requests <#pr>`_.
```
```suggestion
    - Checks that all elements of this document were respected, especially the part about `Pull Requests <#pr>`_.
```

eurunuela replied:
This one should be included.

tsalo commented on docs/contributorfile.rst:
```diff
@@ -132,75 +156,144 @@ Issues labels
+- If there is a Draft PR, you can comment on its development in the message board or making "Comment" reviews. Don't ask for changes, and especially, **don't approve the PR**
+- If the PR graduated from Draft to full PR, check that it follows the sections `Pull requests <#pr>`_ and `Style Guide <#styling>`_ of these guidelines. If not, invite the author to do so before starting a review.
+- **Don't limit your review to the parts that are changed**. Look at the entire file, see if the changes fit well in it, and see if the changes are properly addressed everywhere in the code - in the documentation, in the tests, and in other functions. Sometimes the differences reported don't show the full impact of the PR in the repository!
+- Unless it's for typo fixes or similar, invite the author of the PR to make changes before actually doing them yourself. Request changes via comments or in the message board or by checking out the PR locally, making changes and then submitting a PR to the author's branch.
```
To clarify, does this mean that reviewers shouldn't use the code suggestion tool (like you probably see throughout my review) or that they shouldn't directly commit to the contributor's branch? I personally like the code suggestion tool (obviously), so I'd like to allow reviewers to use it as needed.

smoia replied:
That's a practice I liked a lot and that I tried to pick up from @rmarkello: he never used the suggestion tools, opting for explaining the changes to the code he meant to apply and waiting for the PR author to do them instead - with the exception of rewording or typos or minimal things.
If you don't agree with this practice, we can reword it so that it's only against direct commits. I know that the suggestion tool make things waaaaaaaaaay faster, but for learning purposes is not the best.

eurunuela replied:
Maybe we could make an exception with newbies and make it educational for them. Or if one of us would like an extended description of the suggestion, we can just ask for it!

eurunuela replied:
To me this PR is ready to be merged once we figure this part out.

RayStick replied:
I don't _think_ I have used the code suggestion tool before. I normally just make comments, or put in a PR to the author's branch, with my suggested changes. I guess it would be good to be clear about what we prefer, if anything.

smoia replied:
I'm going to rewrite this point in the next commit!

tsalo commented on docs/contributorfile.rst:
```diff
@@ -132,75 +156,144 @@ Issues labels
+- In case of missing tests or updates to user documentation:
+    - Asks for more documentation or tests before approving the PR, *or*
+    - Checks that the adequate issues have been opened to address the lack of documentation or tests (1 issue per item).
+- Double-checks that the title is clear and the labels are correct to trigger an adequate ``auto`` release - feel free to change them.
```
```suggestion
- Double-checks that the title is clear and the labels are correct to trigger an appropriate ``auto`` release - feel free to change them.
```

tsalo commented on docs/contributorfile.rst:
```diff
@@ -132,75 +156,144 @@ Issues labels
+- Decides what to do in case of a coverage decrease (in *codecov/patch*).
+- In case of missing tests or updates to user documentation:
+    - Asks for more documentation or tests before approving the PR, *or*
+    - Checks that the adequate issues have been opened to address the lack of documentation or tests (1 issue per item).
```
```suggestion
    - Checks that the appropriate issues have been opened to address the lack of documentation or tests (1 issue per item).
```

tsalo commented on docs/contributorfile.rst:
```diff
@@ -132,75 +156,144 @@ Issues labels
+    - Checks that the adequate issues have been opened to address the lack of documentation or tests (1 issue per item).
+- Double-checks that the title is clear and the labels are correct to trigger an adequate ``auto`` release - feel free to change them.
+- **Is the one that is going to merge the PR.**
+- After the PR got merged and a new release was triggered, checks that:
```
```suggestion
- After the PR has been merged and a new release has been triggered, checks that:
```

tsalo commented on docs/contributorfile.rst:
```diff
@@ -132,75 +156,144 @@ Issues labels
 -------------
-To improve understanding pull requests "at a glance", we use the same labels used for issues. Multiple labels can be assigned - just think which ones suit your PR the most! 
-When opening a pull request, assign it to at least one label.
+To improve understanding pull requests "at a glance", we use the labels listed above. Multiple labels can be assigned, but it's a good idea to keep different types of contributions separated, unless they are minimal - for instance, you might want to open a PR for code, one for documentation and one for testing.
```
This confuses me a little. It implies to me that one set of changes (e.g., adding a new metric to `phys2denoise`), would involve three PRs: one to add the metric, one to add documentation about the metric, and one to add tests for the metric. The last one clearly falls under the first PR based on the rest of this document, but I don't think it's clear from context if contributors should separate the PR with the new metric+testing from another PR with documentation about the metric. Personally, I think all of it should be done in one PR in that situation.

smoia replied:
The reasoning is two(three?)fold.
First, if we want to use the automatic `all-contributors` plugin for `auto`, if reviewers use the "suggestion" tool they are interpreted as co-authors of the PR. For instance, this PR is labelled "internal" and "documentation", so all reviewers I accepted suggestion from will get the contribution badge for `documentation` and `infrastructure`. Normally it doesn't matter so much because we tend to review the types of PR that match our previous contribution, but I already have to clean up a bit the contribution table.
Second, since the main focus of a `code` PR is code changes (and not tests or documentations, that have to be done but are _secondary_ in that context), even if the PR contains changes to the tests and documentation I would only label it with one of the two code-related label (majormod and minormod), since the code change is the real scope of the PR. This is also related to the fact that all-contributors has a binary nature (either you contributed in a field or you didn't). I find slightly unfair that two contributors have the same badge (e.g. `test`), but one made almost all the tests and the other two small changes. This is partially solved by the number and size of the PRs labelled in a certain way, so having clear, distinct PRs helps adding a metric of contribution.
Bonus reason: we could _not_ use the all-contributors plugin for `auto`, but most of the time it helps adding contributions impartially and it really helps keeping track of most of the contributions.

eurunuela replied:
Gotcha! Still, I think tests could go with code changes into the same PR. If we make two different PRs we will increase our version with each of them, which does not make sense to me if the PRs are 100% related like code + tests.

RayStick replied:
I think it might make sense to include tests alongside the code, in the same PR, as well. 

And if a contributor writes the code, but is not sure about how to write the tests, that is when they can make a draft PR and ask for help from other people to include tests in their PR?

smoia replied:
I'm going to update after the discussion in the call!

tsalo reviewed (changes requested):
I mostly have nitpicky wording requests that you can take or leave, but there are a couple of questions in there.

eurunuela commented on docs/contributorfile.rst:
```diff
@@ -132,75 +156,144 @@ Issues labels
 Docstrings should follow `numpydoc <https://numpydoc.readthedocs.io/en/latest/format.html>`_ convention. We encourage extensive documentation.
 The python code itself should follow `PEP8 <https://www.python.org/dev/peps/pep-0008/>`_ convention whenever possible: there are continuous integration tests checking that!
-You can use linters to help you write your code following style conventions. Linters are add-ons that you can run on the written script file. We suggest the use of **flake8** for Python 3. Many editors (Atoms, VScode, Sublimetext, ...) support addons for online lintering, which means you’ll see warnings and errors while you write the code - check out if your does!
+You can use linters to help you write your code following style conventions. Linters are add-ons that you can run on the written script file. We suggest the use of **flake8** for Python 3. Many editors (Atom, VScode, Sublimetext, ...) support addons for online lintering, which means you’ll see warnings and errors while you write the code - check out if your does!
```
Should we wait for the discussion going on at #324 before merging?

smoia replied:
Even if we adopt an opinionated code formatter, I wouldn't force the contributors to adopt it during their coding or documentation writing. In fact, one of the reasons to adopt an opinionated code formatter is to _not have to respect any opinionated code format during coding_. Most of the time it's sufficient that the code respects *flake8* conventions, so that the changes introduced by opinionated code formatters do not introduce an incredible amount of differences compared to the submitted code - although it's true that it becomes less essential. Since we're going to have `black` run during PR merge (I guess?), I would keep this section as it is.

eurunuela reviewed (commented):
I'm happy with the PR but think @tsalo 's suggestions are good and should be addressed first.

I also have a question regarding #324

tsalo commented on docs/contributorfile.rst:
```diff
@@ -116,12 +126,26 @@ At physiopy, we use Issues and Milestones to keep track of and organise our work
+
+Issue & PR labels
+~~~~~~~~~~~~~~~~~~~
+-  Documentation: Improvements or additions to documentation. This category includes (but is not limited to) docs pages, docstrings, and code comments.
```
This is a minor thing, but it would be awesome if the labels themselves could be replaced with badges.

tsalo replied:
Here's an example (built [here](https://shields.io/category/issue-tracking)): https://img.shields.io/github/labels/physiopy/phys2bids/Documentation?style=plastic

eurunuela replied:
> This is a minor thing, but it would be awesome if the labels themselves could be replaced with badges.

You mean in the bullet list?

tsalo replied:
Yeah, just in the documentation.

tsalo replied:
And the badges could probably link to the list of issues with each label, if we want.

eurunuela replied:
That would be very convenient actually! I like that.

smoia replied:
Yeah I wondered how did you do it in `tedana`, cause I definitely wanted to do it here!
I wouldn't stall this PR for this reason, but we can definitely open an issue about it!

smoia replied:
(Also I would use the _flat_ style rather than the plastic one)

smoia commented:
Ok, addressed most of the suggestions and questions!
However it might be worth addressing those questions during an upcoming meeting. What do you think?

RayStick commented:
It seems we are waiting to merge this, after discussion in the next meeting on some of the topics raised, is that correct?

smoia commented:
@tsalo @RayStick @eurunuela I updated a couple of sentences to reflect the discussion we had here and during the last meeting.

Please have a look!

eurunuela commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,7 +1,37 @@
+
 
+## Proposed Changes
+<!-- List here major points of changes, so that the reviewers can have a bit more context while looking at your work! -->
```
```suggestion
<!-- List major points of changes here, so that the reviewers can have a bit more context while looking at your work! -->
```

eurunuela commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,7 +1,37 @@
+<!-- If relevant, you can add a milestone label or indicate an ideal timeline for this PR, including whether the progress of this PR is linked to other PRs being completed before or after it. -->
+- [ ] I added information regarding the timeline of completion for this PR.
+<!-- If you want, you can ask for reviews on a draft PR -->
+- [ ] Please, comment on my PR while it's a draft and give me feedback on the development!
```
Shouldn't this last sentence be for the reviewers?

smoia replied:
Ah, no. This is for the user to clarify that they want a review while drafting (otherwise in the documentation we wrote not to do that)

eurunuela commented on docs/contributorfile.rst:
```diff
@@ -78,36 +80,44 @@ If you want to perform one, open an issue on GitHub or drop a comment in Gitter,
+
+Contributing with Pull Requests Reviews
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+One big challenge in this kind of software development is merging code accurately but without waiting too much time.
```
```suggestion
A big challenge of software development is merging code accurately but without having to wait too much time.
```

eurunuela commented on docs/contributorfile.rst:
```diff
@@ -132,75 +156,145 @@ Issues labels
-To improve understanding pull requests "at a glance", we use the same labels used for issues. Multiple labels can be assigned - just think which ones suit your PR the most! 
-When opening a pull request, assign it to at least one label.
+To improve understanding pull requests "at a glance" and use the power of ``auto``, we use the labels listed above. Multiple labels can be assigned to a PR - in fact, all those that you think are relevant.
+We strongly advise to keep the changes you're introducing with your PR limited to your original goal. Little corrections here and there in the code that you're already modifying are a great help, but they risk complicating the nature of the PR and the reviewing process. It is much better to open another PR with the objective of doing such corrections!
```
I don't agree with this. Why would we want to merge some code that we know needs corrections? We should not merge code that contains errors.

Another thing is that from that PR we think of another PR with a different goal. Then it makes sense to work on a new PR after merging.

smoia replied:
Let's see if the new rewording helps clarifying this point (commit to follow).

eurunuela commented on docs/contributorfile.rst:
```diff
@@ -132,75 +156,145 @@ Issues labels
-When opening a pull request, assign it to at least one label.
+To improve understanding pull requests "at a glance" and use the power of ``auto``, we use the labels listed above. Multiple labels can be assigned to a PR - in fact, all those that you think are relevant.
+We strongly advise to keep the changes you're introducing with your PR limited to your original goal. Little corrections here and there in the code that you're already modifying are a great help, but they risk complicating the nature of the PR and the reviewing process. It is much better to open another PR with the objective of doing such corrections!
+In general, if you're tempted to assign more than one label that would trigger a release, you might want to split your PR instead.
+When opening a pull request, assign it at least one label.
```
This is also a bit confusing. So, it's better to split the PR when assigning more than one label. But then you say the PR should have at least one label. It sounds like we're encouraging people to give more than one label with the "at least".

smoia replied:
Same, let's see if it's better now.

eurunuela reviewed (changes requested):
My comments are not critical but I think some of the rephrasing @tsalo suggested should be included.

smoia commented:
@eurunuela do you think it's clearer now?

eurunuela commented on docs/contributorfile.rst:
```diff
@@ -132,75 +156,144 @@ Issues labels
+- Your code respects the `adopted Style <#styling>`_, especially:
+    - Your code is lintered adequately and respects the `PEP8 <https://www.python.org/dev/peps/pep-0008/>`_ convention.
+    - Your docstrings follow the `numpydoc <https://numpydoc.readthedocs.io/en/latest/format.html>`_ convention.
+    - There is no grammar mistake or typo and the text is fluid.
```
I think this one should be inlcuded.

eurunuela commented on docs/contributorfile.rst:
```diff
@@ -132,75 +156,144 @@ Issues labels
+
+Reviewing PRs
+-------------
+Reviewing PRs is an extremely important task in collaborative development. In fact, it is probably the task that requires the most time in the development, and it can be stressful for both the reviewer and the author. Remember that, as a PR Reviewer, you are granting that the codes works and integrates well with the rest of the repository, hence **you are responsible for the quality of the reposi ...
```
I think this one should be included too.

eurunuela commented:
> @eurunuela do you think it's clearer now?

Sorry I didn't make it clearer before. I commented on the ones I think should be included.

smoia commented:
@eurunuela now?

eurunuela reviewed (approved):
LGTM! Thank you!

smoia commented:
@tsalo @RayStick if it's ok for you we can merge this in?

tsalo reviewed (approved):
I'm happy with the changes. Thank you.

smoia commented:
@RayStick do you mind if we merge this?

eurunuela commented:
> @RayStick do you mind if we merge this?

She already approved, so I guess she's happy 🤷🏻‍♂️

eurunuela commented:
We should celebrate this!

RayStick commented:
Great.  :tada:

smoia commented:
<!-- GITHUB_RELEASE COMMENT: released -->
:rocket: PR was released in [`2.3.0`](https://github.com/physiopy/phys2bids/releases/tag/2.3.0) :rocket:
